homeassistant/components/neptune_apex/config_flow.py

```diff
@@ -0,0 +1,88 @@
+"""Config flow for Neptune Apex integration."""
+import logging
+
+from pynepsys import Apex
+import voluptuous as vol
+
+from homeassistant import config_entries, core, exceptions
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
+
+from .const import DOMAIN  # pylint:disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+    }
+)
+
+
+class ApexHub:
+    """Handle authentication validation during config flow."""
+
+    def __init__(self, host):
+        """Initialize."""
+        self.host = host
+        self.apex = None
+
+    async def authenticate(self, username, password) -> bool:
+        """Test if we can authenticate with the host."""
+        self.apex = Apex(self.host, username, password)
+        result = await self.apex.validate_connection()
+        if result == "success":
+            return True
+        if result == "invalid_auth":
+            raise InvalidAuth
+        raise CannotConnect
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    hub = ApexHub(data[CONF_HOST])
+
+    if not await hub.authenticate(data[CONF_USERNAME], data[CONF_PASSWORD]):
+        raise InvalidAuth
```

Comment:
This can never happen since we raise inside `hub.authenticate`.

Reply:
Got it.